source/common/buffer/buffer_impl.h

```diff
@@ -307,6 +314,9 @@ class Slice {
   void transferDrainTrackersTo(Slice& destination) {
     destination.drain_trackers_.splice(destination.drain_trackers_.end(), drain_trackers_);
     ASSERT(drain_trackers_.empty());
+    // The releasor needn't to be transferred, and actually if there is releasor, this
+    // slice can't coalesce. Then there won't be a chance to calling this method.
+    ASSERT(releasor_ == nullptr);
```

Comment:
I might be missing something on my end but how can we guarantee no slice that we're transferring (copying) has no releasor?

Reply:
There is condition control that https://github.com/envoyproxy/envoy/blob/782bb0c102ffbd9880615526c54c00301a6f3f82/source/common/buffer/buffer_impl.cc#L298-L303

The `other_slice.canCoalesce()` is false for the Slice created from BufferFragement.
https://github.com/envoyproxy/envoy/blob/782bb0c102ffbd9880615526c54c00301a6f3f82/source/common/buffer/buffer_impl.h#L89-L90

Reply:
That makes sense, thanks @soulxu